abstract-spec/smt-spec/ssf.smt

```diff
@@ -0,0 +1,329 @@
+;
+; Thomas Pani, 2024
+;
+; Run with
+;   $ cvc5 ssf.smt
+;
+; Tested with CVC5 v1.2.0.
+
+(set-logic ALL)
+
+(set-option :produce-models true)
+; we need finite model finding to answer sat problems with universal
+; quantified formulas
+(set-option :finite-model-find true)
+; enable extended set operators
+(set-option :sets-ext true)
+; we need to enable incremental mode to answer multiple queries
+(set-option :incremental true)
+
+; ============================
+; Datatypes
+; ============================
+
+; we search a finite universe: 3 hashes, 5 checkpoints, 4 nodes
+(declare-datatype Hash ((Hash1) (Hash2) (Hash3)))
+(declare-datatype Checkpoint ((C1) (C2) (C3) (C4) (C5)))
+(declare-datatype Node ((Alice) (Bob) (Charlie) (David)))
+(define-fun N () Int 4)
+(declare-datatype Vote ((Vote (source Checkpoint) (target Checkpoint) (sender Node))))
+
+; ============================
+; Blocks
+; ============================
+
+; genesis block
+(declare-const genesis Hash)
+
+; edges, genesis has a self-loop
+(declare-fun parent_of (Hash) Hash)
+(assert (= (parent_of genesis) genesis))
+
+; slots
+(declare-fun slot (Hash) Int)
+(assert (= (slot genesis) 0))
+
+; slots are increasing from parent to child
+(assert (forall ((h Hash)) (=> (not (= h genesis)) (> (slot h) (slot (parent_of h))))))
+
+; ancestor-descendant relationship
+(declare-const ancestor_descendant_relationship (Relation Hash Hash))
+(assert (forall ((ancestor Hash) (descendant Hash))
+    (=  (set.member (tuple ancestor descendant) ancestor_descendant_relationship)
+        (or
+            (= ancestor descendant)
+            (and
+                (not (= descendant genesis))
+                (set.member (tuple ancestor (parent_of descendant)) ancestor_descendant_relationship)
+            )
+        )
+    )
+))
+
+; conflicting blocks
+(declare-const conflicting_blocks (Relation Hash Hash))
+(assert (forall ((h1 Hash) (h2 Hash))
+    (=  (set.member (tuple h1 h2) conflicting_blocks)
+        (and
+            (not (set.member (tuple h1 h2) ancestor_descendant_relationship))
+            (not (set.member (tuple h2 h1) ancestor_descendant_relationship))
+        )
+    )
+))
+
+; ============================
+; Checkpoints
+; ============================
+
+; genesis checkpoint
+(declare-const genesis_checkpoint Checkpoint)
+
+; checkpoint block
+(declare-fun checkpoint_block (Checkpoint) Hash)
+(assert (= (checkpoint_block genesis_checkpoint) genesis))
+
+; checkpoint slot
+(declare-fun checkpoint_slot (Checkpoint) Int)
+(assert (= (checkpoint_slot genesis_checkpoint) 0))
+
+; all checkpoints ((B, block_slot), chkp_slot) except genesis have chkp_slot > block_slot
+(assert (forall ((c Checkpoint)) (or (= c genesis_checkpoint) (> (checkpoint_slot c) (slot (checkpoint_block c))))))
+
+; checkpoints have no "identity": if they have the same block and slot, they are the same checkpoint
+(assert (forall ((c1 Checkpoint) (c2 Checkpoint))
+    (=>
```

Comment:
should not that be `=`, that is, "if and only if"?

Reply:
It's not wrong, but the `<=` direction should be implied by EUF? (`checkpoint_block`/`checkpoint_slot` are functions)